src/Couplers/ICICoupler.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef ICICOUPLER_H
+#define ICICOUPLER_H
+
+#include <vector>
+#include "Coupler.hpp"
+#include "../utils.hpp"
+#include "../VectorView.hpp"
+#include <cstring>
+#include <chrono>
+#include <map>
+
+
+using namespace dealii;
+// Class responsible for the implementation of the ionic current interpolation approach
+template<int N_ion>
+class ICICoupler : public Coupler<N_ion> {
+    static constexpr unsigned int dim = 3;
+
+public:
+    // In this case the solveOde method is identical to the one in SVICoupler.
+    // Also in this case the gating variables are defined on dofs.
+    void solveOde(Solver<N_ion>& solver) override {
+        std::shared_ptr<IonicModel<N_ion>> ionicModel = solver.getIonicModel();
+        std::vector<VectorView<TrilinosWrappers::MPI::Vector>> gate_vars_views;
+        for(int i = 0; i < N_ion; i++) {
+            auto [first, last] = gate_vars_owned[i].local_range();
+            gate_vars_views.emplace_back(gate_vars_owned[i], first, last - first);
+        }
+        auto [first, last] = solver.getFESolutionOwned().local_range();
```

Comment:
Beware that this works as long as the partitioning of the vector is contiguous (i.e. rank 0 owns the first n_0 elements, rank 1 the second n_1, and so on and so forth). Actually, Trilinos does **not** make this assumption. Therefore, it might in principle be possible that this fails in quite a hard-to-debug way.

In this case, it would probably be best to work with the `IndexSet` of owned DoFs (the `IndexSet` class is designed basically for this purpose).

This also applies elsewhere.